service/controller/resource/instance/create_scale_workers.go

```diff
@@ -4,126 +4,120 @@ import (
-	cr, err := key.ToCustomResource(obj)
+	azureMachinePool, err := key.ToAzureMachinePool(obj)
 	if err != nil {
-		return "", microerror.Mask(err)
+		return DeploymentUninitialized, microerror.Mask(err)
 	}
 
-	desiredWorkerCount := int64(key.WorkerCount(cr) * 2)
-	r.Logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("The desired number of workers is: %d", desiredWorkerCount))
+	machinePool, err := r.getOwnerMachinePool(ctx, azureMachinePool.ObjectMeta)
+	if err != nil {
+		return DeploymentUninitialized, microerror.Mask(err)
+	}
 
-	currentWorkerCount, err := r.getInstancesCount(ctx, cr, key.WorkerVMSSName)
+	cluster, err := util.GetClusterFromMetadata(ctx, r.CtrlClient, azureMachinePool.ObjectMeta)
 	if err != nil {
-		return "", microerror.Mask(err)
+		return DeploymentUninitialized, microerror.Mask(err)
 	}
-	r.Logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("The current number of workers is: %d", currentWorkerCount))
 
-	allReady, err := vmsscheck.InstancesAreRunning(ctx, r.Logger, key.ResourceGroupName(cr), key.WorkerVMSSName(cr))
+	credentialSecret, err := r.getCredentialSecret(ctx, *cluster)
+	if err != nil {
+		return DeploymentUninitialized, microerror.Mask(err)
+	}
+
+	virtualMachineScaleSetsClient, err := r.ClientFactory.GetVirtualMachineScaleSetsClient(credentialSecret.Namespace, credentialSecret.Name)
+	if err != nil {
+		return currentState, microerror.Mask(err)
+	}
+
+	allReady, err := vmsscheck.InstancesAreRunning(ctx, r.Logger, key.ClusterID(&azureMachinePool), key.NodePoolVMSSName(&azureMachinePool))
 	if vmsscheck.IsVMSSUnsafeError(err) {
 		// VMSS rate limits are not safe, let's wait for next reconciliation loop.
-		return ScaleUpWorkerVMSS, nil
+		return currentState, nil
 	} else if err != nil {
-		return "", microerror.Mask(err)
+		return DeploymentUninitialized, microerror.Mask(err)
 	}
 	// Not all workers are Running in Azure, wait for next reconciliation loop.
 	if !allReady {
-		return ScaleUpWorkerVMSS, nil
+		return currentState, nil
 	}
 
 	// All workers ready, we can scale up if needed.
+	desiredWorkerCount := int64(*machinePool.Spec.Replicas * 2)
+	r.Logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("The desired number of workers is: %d", desiredWorkerCount))
+
+	currentWorkerCount, err := r.GetInstancesCount(ctx, virtualMachineScaleSetsClient, key.ClusterID(&azureMachinePool), key.NodePoolVMSSName(&azureMachinePool))
+	if err != nil {
+		return DeploymentUninitialized, microerror.Mask(err)
+	}
+	r.Logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("The current number of workers is: %d", currentWorkerCount))
+
 	if desiredWorkerCount > currentWorkerCount {
-		// Raise the worker count by one
-		err = r.scaleVMSS(ctx, cr, key.WorkerVMSSName, currentWorkerCount+1)
+		err = r.ScaleVMSS(ctx, virtualMachineScaleSetsClient, key.ClusterID(&azureMachinePool), key.NodePoolVMSSName(&azureMachinePool), int32(desiredWorkerCount))
```

Comment:
let's try scaling straight to the desired number of workers, instead of going one by one.